### DISCUSSION TITLE
Proposal: special syntax for lambdas as last arguments in invocation expressions (for DSLs)

### DISCUSSION THREAD

**Discussion by orthoxerox** [dotnet/csharplang#1151] [General]
Warning: lots of Kotlin-inspired sugar.

Rationale: C# as it is doesn't support DSLs that well. Several features in Ruby, Groovy and Kotlin allow developers to author code that looks like native constructs.

1. Lambdas with receivers.

Lambdas can have their first parameter be treated as an implicit receiver inside the body. This affects name resolution in lambda body: implicit references to `this` of the declaring method are not available inside the body of a lambda. `@` is used as an alias of the receiver parameter, `this` is used to refer to the receiver of the declaring method and all symbols not resolved as lambda locals are resolved as implicit lambda receiver members before other options are tried.

Methods indicate they accept lambdas with receivers by prefixing the parameter type with `this`, e.g., `Action<this Customer>` or `Func<this Customer, bool>` (this is later compiled as an attribute on the parameter). Such types cannot be used as types of locals, fields, return values or properties.

Lambda declarations in positions where an implicit receiver lambda is expected omit this parameter altogether (i.e., single-parameter lambdas look like zero-parameter lambdas).

Such lambdas are not very useful by themselves, but serve as a stepping stone for the second change.

2. Lambdas as last arguments of an invocation expression.

Zero parameter lambdas (including single parameter lambdas with implicit receivers) passed as the last parameter to a method call can be written outside the parentheses without `() =>`. If this invocation expression is being parsed as an expression statement then the semicolon is optional as well.

When zero parameter lambdas are passed as any other parameter, they can still be written outside the parentheses without `() =>` as long as all other parameters are explicitly named.

```c#
someMethod(arg1, arg2, x => {
    statement(x);
    x.statement;
});

//if x is marked as an implicit receiver
someMethod(arg1, arg2) {
    statement(@);
    statement; //or @.statement;
}

Task.Run(() => {
    Console.WriteLine("Starting");
    DoStuff();
    Console.WriteLine("Done");
}, token);

Task.Run(cancellationToken: token) {
    Console.WriteLine("Starting");
    DoStuff();
    Console.WriteLine("Done");
}
```

Return statements are disallowed in such lambda bodies to prevent confusion.

Expression-bodied lambdas must have their bodies wrapped in `{}` as well. If they are the only argument, then the parentheses are optional as well.:

```c#
f(xs.Select(x => x + 2));

f(xs.Select { @ + 2 });
```

Examples:

```c#
//C# 7.2
var winners = customers
    .Where(c => c.Status == CustomerStatus.VIP)
    .SelectMany(c => db.GetOrders(c))
    .Where(o => o.TotalPrice.AsUSD() > 10000)
    .Shuffle()
    .Take(10);

//c# vNext
//We get shorthand lambdas for free!
var winners = customers
    .Where { Status == CustomerStatus.VIP }
    .SelectMany { db.GetOrders(@) }
    .Where { TotalPrice.AsUSD() > 10000 }
    .Shuffle()
    .Take(10);
```

```c#
//C# 7.2
Task("Clean")
    .Does(() =>
{
    CleanDirectory(buildDir);
});

Task("Restore-NuGet-Packages")
    .IsDependentOn("Clean")
    .Does(() =>
{
    NuGetRestore("./src/Example.sln");
});

Task("Build")
    .IsDependentOn("Restore-NuGet-Packages")
    .Does(() =>
{
    if(IsRunningOnWindows())
    {
      // Use MSBuild
      MSBuild("./src/Example.sln", settings =>
        settings.SetConfiguration(configuration));
    }
    else
    {
      // Use XBuild
      XBuild("./src/Example.sln", settings =>
        settings.SetConfiguration(configuration));
    }
});

Task("Run-Unit-Tests")
    .IsDependentOn("Build")
    .Does(() =>
{
    NUnit3("./src/**/bin/" + configuration + "/*.Tests.dll", new NUnit3Settings {
        NoResults = true
        });
});

//C# vNext
//NB: This is not 100% equivalent to the code above.
//    Task(string) method has been replaced with
//    Task(string, Action<this TaskBuilder> action)
Task("Clean") {
    Does {
        CleanDirectory(buildDir);
    }
}

Task("Restore-NuGet-Packages") 
{
    IsDependentOn("Clean");
    Does 
    {
        NuGetRestore("./src/Example.sln");
    }
}

Task("Build") 
{
    IsDependentOn("Restore-NuGet-Packages");
    Does 
    {
        if(IsRunningOnWindows())
        {
          // Use MSBuild
          MSBuild("./src/Example.sln")
          {
            Configuration = configuration;
          }
        }
        else
        {
          // Use XBuild
          XBuild("./src/Example.sln")
          {
            Configuration = configuration;
          }
        }
    }
}

Task("Run-Unit-Tests") 
{
    IsDependentOn("Build");
    Does 
    {
      NUnit3("./src/**/bin/" + configuration + "/*.Tests.dll", new NUnit3Settings {
        NoResults = true //I wonder if object initializers could get the same treatment
        });
    }
}
```

**Comment by jnm2** [comment]
I do wish C# was more concise, often in Cake but not just in Cake. The whole necessity of passing `new NameOfTheCurrentCommandSettings { Foo = 42 }` to avoid breaking changes and the `.Does(() => { })` lambdas.
I don't know; still thinking.

**Comment by HaloFour** [comment]
> f(xs.Select { x + 2 });

> .Where { Status == CustomerStatus.VIP }

What about this proposal differentiates the very different behavior between these two statements?  Why is `x` the lambda parameter but `Status` is a property of some implicit `this` lambda parameter?

I really don't like the idea of the meaning of `this` being hijacked in this manner.  It feels very JavaScript, and in a very not-good way.

**Comment by MkazemAkhgary** [comment]
```
string Status = "a";
public void Method()
{
    string Status = "b";
    var query = customers.Where { Status == CustomerStatus.VIP };
}
```

so here `Status` in `Where { Status == CustomerStatus.VIP };`
 refers to `x.Status`, what if you wanted to access either of local or global variables in that context?

**Comment by orthoxerox** [comment]
@HaloFour typo, fixed

**Comment by iam3yal** [comment]
In Kotlin the following:
```
someMethod(arg1, arg2, x => {
    statement(x);
    x.statement;
});
```
Would be something like this:
```
someMethod(arg1, arg2, {
    statement(it.x);
    it.x.statement;
});
```
I really like the way Kotlin does things, it reduces code clutter quite a bit, thus make the code easier to read but then again it's a newer language, they probably learnt a lot from C# and other languages.

Personally, I would really like to have something like this:
```
someMethod(arg1, arg2, {
    statement(_.x);
    _.x.statement;
});
```
But I wouldn't mind following Kotlin and have `it` iff it would be feasible as a contextual keyword.

**Reply by tmteam** [reply]
Actually it will be even simpler in kotlin:

```
Kotlin:

someMethod(arg1, arg2) {
    statement(it.x)
    it.x.statement
}

```

And it becomes more valuable when there is nesting:

```
Kotlin:

someMethod(arg1, arg2) { it->
    statement(it.x)
    it.doSome { options->
         println(options)
         println("end")
    }

}


C#:

someMethod(arg1, arg2, x => {
    statement(x);
    x.DoSome( options=>{
         WriteLine(options)
         WriteLine("end")
    });
});

```

**Comment by iam3yal** [comment]
@orthoxerox I still think you have typo there.

The following:
```
.Where { Status == CustomerStatus.VIP }
```
Should be
```
.Where { this.Status == CustomerStatus.VIP }
```
Isn't?

**Comment by orthoxerox** [comment]
@eyalsk No, `Status` is implicitly resolved to be `this.Status`.

**Comment by MkazemAkhgary** [comment]
I think this can be achieved by #91. I'm not trying to be negative here, but such implicit reference can be ambiguous with other local or global literals. 

`.Where(@.Status == CustomerStatus.VIP)`

this also doesn't require any change in method signature.

**Comment by iam3yal** [comment]
@orthoxerox Okay, interestingly confusing. 😆 

How it's going to work with say intellisense?

**Comment by orthoxerox** [comment]
@eyalsk How does it work in method bodies?

@MkazemAkhgary shorthand lambdas are just a bonus that falls out of these changes, the main goal is to be able to write "while-like" or "using-like" custom commands.

**Comment by orthoxerox** [comment]
@HaloFour I wouldn't have hijacked `this` this way, but it's important to be able to introduce a special scope inside the lambda to correctly resolve names valid in this scope without polluting the outer scope or resorting to prefixes. This can be done via a pseudo-receiver, and `this` naturally appears as a consequence. It would've been even stranger if you could say `Foo()` and see that it resolves to `somethingImplicit.Foo()`, but were not able to refer to `somethingImplicit` in any way.

**Comment by iam3yal** [comment]
@orthoxerox What's going to happen in this case? I mean, how do you access a member of the class with the same name within the body of the lambda? or you expect that in this case people will use the traditional syntax?

```
class Foo {
	public int Status { get; set; }

	public void Bar()
	{
		var winners = customers
			.Where { Status == ????.Status }
		    .SelectMany { db.GetOrders(this) }
		    .Where { TotalPrice.AsUSD() > 10000 }
		    .Shuffle()
			.Take(10);
	}
}
```

**Reply by miyu** [reply]
Capture `var self = this` to a local worst case, then use `self` in the closure. It's an edge case but really not that bad.

**Comment by orthoxerox** [comment]
@eyalsk yes, I would expect people to use the traditional syntax.

**Comment by iam3yal** [comment]
@orthoxerox Okay.

In general, I think the idea is good and i really like to get rid of the "ceremony" of lambdas and even more in generics and other places in C# but I do find the case where all members are "exported" and implicitly available in the body confusing, I don't know, maybe it's just a matter of getting used to.. but it may also be the case of taking a good idea a bit too far.

**Comment by orthoxerox** [comment]
@eyalsk the proposal is less about shorthand lambdas and more about providing a way to write constructs like

```c#
Foo("foo") {
    Bar();
    Baz();
    Quux();
}
```

where Bar, Baz and Quux are available and visible only inside that Foo block.

Shorthand lambdas are just a side effect of adding features to the language to support the other use case.

**Comment by iam3yal** [comment]
@orthoxerox Yes but not sure where I said anything about "shorthand lambdas" and your construct is a lambda without the "ceremony".

**Comment by yaakov-h** [comment]
I like the general idea, but I'm not sold on the implicit name resolution or hijacking of `this`.

In particular, the implicit name resolution in nested trailing lambdas (like the build scripts above) would be a nightmare - and which `this` would be which when nested? How can you refer to an outer scope?

**Comment by bondsbw** [comment]
Agreed with the comments against hijacking `this`, and agreed with @eyalsk:
>But I wouldn't mind following Kotlin and have `it` iff it would be feasible as a contextual keyword.

Then:
- `it` always resolves to the context of the lambda
- `this` always resolves to the context of the caller (only available in non-static contexts)
- `it` can be implied in such a lambda, but `this` cannot

@eyalsk's code becomes:
```c#
class Foo {
    public int Status { get; set; }

    public void Bar()
    {
        var winners = customers
            .Where { Status == this.Status } // `it` is implied on first Status; `this` is required on second
            .SelectMany { db.GetOrders(it) } // or simply .SelectMany(db.GetOrders)
            .Where { TotalPrice.AsUSD() > 10000 }
            .Shuffle()
            .Take(10);
    }
}
```

**Comment by CyrusNajmabadi** [comment]
I'm *slightly* amenable to all this.   :)

**Comment by bondsbw** [comment]
From the proposal:
> Return statements are disallowed in such lambda bodies to prevent confusion.

I don't think this is a necessary restriction.  It shouldn't be any more confusing than in a regular lambda or local function.

**Comment by CyrusNajmabadi** [comment]
The biggest thing for me is that i'm concerned in practice that you'll still end up needing the lambda parameter to be named, and this will fall flat.  especially in the deeply nested lambda case.  In my deep lambdas i'm going to want to compare the deep 'it' to a higher level 'it', and i'm going to be SOL.   if i could still write a lambda parameter such that this was ok:

```c#
Task("Clean") t1 => {
    Does t2 => {
        // Use t1 and t2 here
        CleanDirectory(buildDir);
    }
}
```

Then that might make me feel less uneasy about all this.  Of course, if you didn't need the parameters, then you could elide them.

**Comment by CyrusNajmabadi** [comment]
Honestly, things like this do resonate with me:

```c#
Task("Restore-NuGet-Packages") 
{
    IsDependentOn("Clean");
    Does 
    {
        NuGetRestore("./src/Example.sln");
    }
}
```

Clean, readable (presuambly not syntactically ambiguous).  Not too magical (once you learn the rules about lambdas + plus final arguments).  

It fits core desires that led to our existing lambda syntax in the first place.  Namely: reducing ceremony, allowing user logic to shine, being expressive enough to handle important use cases, etc. etc.

**Comment by CyrusNajmabadi** [comment]
The only downside, afaict, is that the ship has sailed with existing APIs.  For example, this is a language feature that *by design* is intertwined with API design.  It's effectively saying: your delegate parameters should go last.

We've already shipped tons of APIs that don't follow that pattern.  And it conflicts with existing API guidance around putting things like CancellationTokens last.  For example, Task.Factory.StartNew() places the delegate first.  That's super unfortunate.  Similarly, Roslyn itself has lots of APIs that don't put the delegate last because that's where the cancellation token goes.  :-/

**Comment by bondsbw** [comment]
@CyrusNajmabadi How about
```c#
Task.Factory.StartNew(cancellationToken: token)
    {
        ...
    };
```
or
```c#
Task.Factory.StartNew(token)
    action: {
        ...
    };
```

**Comment by CyrusNajmabadi** [comment]
maybe :)  

warming up again :D

**Comment by CyrusNajmabadi** [comment]
if we do that, we def can't ever make semicolons optional BTW :D

**Comment by bondsbw** [comment]
Though I suspect many third-party fluent APIs would just add overloads.

**Comment by CyrusNajmabadi** [comment]
To help with the proposal, i would love to see some other examples of existing fluent APIs converted over to this form.  Maybe 1-2 examples of real world code using well known APIs.  Showing the before/after would be nice.  Thanks!

**Comment by CyrusNajmabadi** [comment]
can't :)  They're reserved names.  you could always do ```@if (a) { }```, and that would be totally ok and unambiguous.

**Comment by yaakov-h** [comment]
I realised that just after I posted it, so deleted it.

You're too fast 😄

**Comment by orthoxerox** [comment]
As I have expected, various builders are probably the best looking showcases for this new syntax. I'll add ASP .Net MVC HTML helpers later. They should look good as well.

```c#
//NUnit
Assert.That(phrase, Does.Contain("World"));
Assert.That(phrase, Does.Not.Contain("goodbye"));
Assert.That(phrase, Does.Contain("WORLD").IgnoreCase);
Assert.That(phrase, Does.Not.Contain("BYE").IgnoreCase);

assert(phrase) {
    Does.Contain("World");
    Does.Not.Contain("goodbye");
    Does.Contain("WORLD").IgnoreCase;
    Does.Not.Contain("BYE").IgnoreCase;
}

//language-ext
//(I went back and forth about allowing multiple trailing lambdas
// when writing the OP, so the syntax here is rough and not though-out at all,
// the real match expression might obsolete it altogether
return (await program()).Match(
	Some: gender => new Result<Gender>(gender),
	None: () => Result<Gender>.None,
	Fail: e => Result<Gender>.Bottom
);

return match(await program())
	Some { new Result<Gender>(this) }
	None { Result<Gender>.None }
	Fail { Result<Gender>.Bottom };

//Newtonsoft.Json
JObject rss =
    new JObject(
        new JProperty("channel",
            new JObject(
                new JProperty("title", "James Newton-King"),
                new JProperty("link", "http://james.newtonking.com"),
                new JProperty("description", "James Newton-King's blog."),
                new JProperty("item",
                    new JArray(
                        from p in posts
                        orderby p.Title
                        select new JObject(
                            new JProperty("title", p.Title),
                            new JProperty("description", p.Description),
                            new JProperty("link", p.Link),
                            new JProperty("category",
                                new JArray(
                                    from c in p.Categories
                                    select new JValue(c)))))))));

JObject rss = obj {
	prop("channel") {
		prop("title", "James Newton-King");
		prop("link", "http://james.newtonking.com");
		prop("description", "James Newton-King's blog.");
		prop("item") {
			arr {
				from p in posts
				orderby p.Title
				select obj {
					prop("title", p.Title);
					prop("description", p.Description);
					prop("link", p.Link);
					prop("category") {
						arr {
							from c in p.Categories
							select new JValue(c)
						}
					}
				}
			}
		}
	}
}
```

**Comment by CyrusNajmabadi** [comment]
One more ting.  I *highly* doubt there would be an accepted form that did implicit member lookup.  so i think we'll have to have at least  ```@.``` ```_.``` or ```it.```.  Can you amend the examples appropriately?

**Comment by CyrusNajmabadi** [comment]
Note: the JObject case is actually *uncompelling* to me.  It saves typing 'new'.  but has a super ugly train of ```}```s at the end.   And if ```new``` is a problem, you could just have have helpers like so:

```c#
static class J {
    public static JProperty Property(...) => new JProperty(...);
}
```

so you could then write:

```c#
JObject rss =
    J.Object(
        J.Property("channel",
```

**Comment by svick** [comment]
@CyrusNajmabadi And you don't even need the `J.` if you use `using static`.

Though that makes me wonder: @orthoxerox, how do the outermost singe-identifier calls (e.g. `assert(phrase) { … }`) work? Are you assuming some `using static`?

**Comment by theunrepentantgeek** [comment]
Similar to my proposal of Block Parameters (see #589), though that wasn't inspired by Kotlin. 

Instead of overriding **this**, that proposal introduces the new keyword **block**.

**Comment by orthoxerox** [comment]
@svick Yes, `using static`. Kotlin has standalone static functions that are used to create the outermost builder, the C# way is `using static`.

@CyrusNajmabadi without implicit member lookup the builder/DSL use case loses a lot of attractiveness, since you either prefix almost everything inside with the chosen sigil or lose type safety. The proposal will then become just another syntax for shorthand lambdas (#91).

One thing I can suggest is that such lambdas can still hijack implicit member lookup, but not `this`. Instead of `this` we can use `@`, for example, and write `xs.Where{Status == this.Status}` or `xs.Where{@.Status == this.Status}` if total unambiguity is desired. This, of course, doesn't resolve the issue of unavoidable name shadowing of outer locals by members of `@`, but name clashes between locals and members can be avoided by following C# naming guidelines.

**Comment by jveselka** [comment]
I concur that implicit member lookup would be very benefical if not necessary for this feature to be really usefull. However I would suggest using `implicit` keyword (its perfect match and `this` implies extension methods)
`Func<implicit Customer, bool>`
It should be available also for regular delegate parameters
`delegate bool CustomerPredicate(implicit Customer c);`
And if it goes that far I see no reason why it shouldn't be general language construct available even in regular method parameterss.
```C#
void FillList(implicit List<string> list)
{
  Add("some");
  Add("items");
  Sort();
}
```
But that should probably be separate proposal.

**Comment by GeirGrusom** [comment]
I have some use cases.

```csharp
// Mock setups
// Moq
foo.Setup { it.DoSomething() }.Returns(true);

// NSubstitute
foo.WhenForAnyArgs { it.DoSomething }.Return(true);
foo.DoSomething(out int bar).Then { it[0] = 100 };

// Code timing
void TimeBlock(Action handle)
{
  var stp = Stopwatch.StartNew();
  handle();
  Log($"Whatever took {stp.TotalElapsedMilliseconds} ms");
}

TimeBlock
{
  CalculatePi();
}

// Custom IDisposable pattern
void UsingClosable<TClosable>(TClosable subject, Action<TClosable> action) 
    where TClosable : IClosable
{
  try
  {
    action(subject);
  }
  finally
  {
    subject.Close();
  }
}

UsingClosable(GetClosable())
{
  it.WriteFoo();
}
```

**Comment by iam3yal** [comment]
Here is another use-case written in Kotlin:

```
disposable = DownloadServiceObservable(source, destination)
                .download()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext{
                    val percentage = calculatePercentage(it.receivedLength, it.totalLength)

                    progressbar.progress = percentage
                    counter.text = "$percentage%"
                }
                .doOnComplete{
                    video.setOnPreparedListener{ video.start() }
                    video.setVideoPath(path)
                }
                .doOnTerminate{
                    source.close()
                    destination.close()
                }
                .subscribe()
```
C# equivalent would be something like this:
```
disposable = DownloadServiceObservable(source, destination)
                .Download()
                .SubscribeOn(Schedulers.IO())
                .ObserveOn(AndroidSchedulers.MainThread())
                .DoOnNext(it => {
                    var percentage = CalculatePercentage(it.ReceivedLength, it.TotalLength)

                    progressbar.Progress = percentage
                    counter.Text = $"{percentage}%"
                })
                .DoOnComplete(() => {
                    video.SetOnPreparedListener(() => video.Start())
                    video.SetVideoPath(path)
                })
                .DoOnTerminate(() => {
                    source.Close()
                    destination.Close()
                })
                .Subscribe()
```

**Comment by Happypig375** [comment]
Regarding the potential new keyword for referring to the implicit lambda parameter, I would add a new option: `me`. I think `it` sounds better, but `me.Property` looks better than `it.Property`.

P.S. Wow, 40 comments in the first hour! I am very surprised and impressed. Never seen any issues/PRs this hot (in any repo, not just csharplang), definitely a record. Usually issues in csharplang recieve only one or two comments in the first hour.

**Comment by orthoxerox** [comment]
@Happypig375 you should check out the default interface methods issue. Or the leaky out vars decision in dotnet/roslyn. Now that were some hot discussions.

**Comment by Happypig375** [comment]
At least I witnessed this issue, I was not there in the first hour of those two. :disappointed:

**Comment by MkazemAkhgary** [comment]
>P.S. Wow, 40 comments in the first hour! I am very surprised and impressed. Never seen any issues/PRs this hot (in any repo, not just csharplang)

@Happypig375 you probably missed #993

**Comment by Happypig375** [comment]
Hopefully no more unconstructive flame wars like that in the future :wink:

**Comment by quinmars** [comment]
@GeirGrusom, The time block use case can be accomplished with a using block, which supports some essential C# features:

```csharp
static public IDisposable TimeBlock() => new TimeBlockDisposable { Watch = Stopwatch.StartNew() };
class TimeBlockDisposable : IDisposable
{
    public Stopwatch Watch;
    public void Dispose() => Debug.WriteLine($"Whatever took {Watch.ElapsedMilliseconds} ms");
}

public static IEnumerable<double> Pi()
{
    using (TimeBlock())
    {
        // var pi = CalculatePi();

        // The nice thing is we could also
        // test our async algorithm here.
        // var pi_async = await CalculatePiAsync();

        // Or we could return all intermediate
        // results.
        double pi = 3.14;
        while (true)
        {
            pi = CalculatePiIteratively(pi);
            yield return pi;
        }
    }
}
```
The same can be done for the custom `IDisposable` pattern.

**Comment by orthoxerox** [comment]
Updated the OP to exclude `this` shadowing.

**Comment by gafter** [comment]
I wrote and spoke publicly about this kind of *control abstraction* that I had [proposed and implemented for Java around 2006-2008](http://openjdk.java.net/challenge/).

**Comment by orthoxerox** [comment]
@gafter How did your implementation deal with `return` statements?

**Comment by gafter** [comment]
Exception-handling (throwing and catching exceptions) is exceptionally efficient in the Java runtime. Creating exceptions not so efficient because it captures a stack trace. My implementation allocates an exception object lazily, when needed, without capturing a stack trace, and throws it, catching it in the enclosing construct where the `return` statement semantics are executed.

This can even be made to work in concurrent control constructs, where the exception needs to be shuttled from one thread to another. See http://gafter.blogspot.com/2006/10/concurrent-loops-using-java-closures.html . I believe the prototype contains an implementation of this concurrent loop construct.

**Comment by gafter** [comment]
In case that did not answer your question, a `return` statement returns from the enclosing method, not the implicit lambda. Similarly `break` and `continue` work "as expected".

**Comment by orthoxerox** [comment]
@gafter Thanks, your answer was very clear. So basically the code would be lowered to something like this, wouldn't it?

```c#
foo(bar) {
    return 1;
    break;
    continue;
}
```

```c#
try {
    foo(bar, () => {
        throw ControlFlowExceptionCache.Return(1);
        throw ControlFlowExceptionCache.Break();
        throw ControlFlowExceptionCache.Continue();
    });
}
catch (ReturnException<int> ex) {
    return ex.ReturnValue;
}
catch (BreakException ex) {
    break;
}
catch (ContinueException ex) {
    continue;
}
```

Do you happen to know why JetBrains didn't do that in Kotlin?

**Comment by gafter** [comment]
Something like this http://gafter.blogspot.com/2006/09/control-abstraction.html

**Comment by orthoxerox** [comment]
@gafter Yes, but with the option to change the default receiver inside the block, so you can say

```
Panel {
    Color = #FF0000;
    Button {
        Text = "Press me"
    }    
}
```

**Comment by rollsch** [comment]
Any updates on this? Looks very neat.

**Comment by lucasteles** [comment]
That would be very good for MAUI

One of the premises is to have declarative UI that end up needing some kind of DSL

This proposal could help a lot

**Comment by CyrusNajmabadi** [comment]
This is a mythical man-month issue.  You can't just throw more teams at it hoping it will go faster :-/

**Comment by CyrusNajmabadi** [comment]
> Any updates on this? Looks very neat.

There are no updates on this.  This would need to be championed by someone to move anywhere.

**Comment by marchy** [comment]
Is this on the consideration list for C# 11?

The RemObjects C# compiler supports this [Trailing Closures](https://docs.elementscompiler.com/Hydrogene/LanguageExtensions/TrailingClosures/) feature quite nicely, and Swift and Kotlin have supported this from the get-go (5+ years now). Would be great to see C# catch up to this modern must-have that can unlock all sorts of DSL-style APIs.

**Reply by fa1ntl1ght** [reply]
+1 for this, looks very useful in Kotlin

**Reply by marchy** [reply]
Posted a more thorough set of real-world examples on related issue https://github.com/dotnet/csharplang/issues/2781.
Not sure which should be the active one so linking both.

**Reply by 333fred** [reply]
Neither is an active issue championed by a member of the LDM. I do have some interest in it, and likely would champion if someone wanted to write up a specification for it. It would definitely not meet the bar for C# 11, however. That would have needed to happen this time last year.

**Reply by marchy** [reply]
@333fred are you aware of any .NET frameworks/APIs that could benefit from this?

Perhaps we could start itemizing some of these to help drive momentum, similar to how the EF team is tracking [features blocked on lack of support for modern expression tree constructs](https://github.com/dotnet/efcore/issues/27047).

No idea what getting a spec spun up would entail.
Are you referring to inside or outside MS?

**Reply by 333fred** [reply]
I am not aware immediately.

A spec can be written by anyone. There is an issue template that covers it.

**Reply by orthoxerox** [reply]
> if someone wanted to write up a specification for it

If someone *wrote* up a specification 😛, I don't doubt many people want to, but won't. I'll see what I can come up with.

**Comment by marchy** [comment]
Here are a handful of examples sweeping through our codebase to kick things off:

### Frameworks Improved

## LINQ
**Example: All `Select` / `Where` etc. operators**
```
IEnumerable<Thing> things = (await eventsQuery
	.Select( _ => new {
		_.ID,
		_.Name,
		_.DALPlace,
	}).ToListAsync()
).Select( _ => new {
	_.ID,
	_.Name,
	Place = JsonConvert.DeserializeObject<PlaceDBO>( _.DALPlace ),
});
```
⇨ ⇨ ⇨
```
IEnumerable<Thing> things = (await eventsQuery.Select {
	new {
		_.ID,
		_.Name,
		_.DALPlace,
	}
}.ToListAsync()).Select {
	new {
		_.ID,
		_.Name,
		Place = JsonConvert.DeserializeObject<PlaceDBO>( _.DALPlace ),
	}
};
```

## Task API
**Example: Continuations**
```
return UnitOfWork.SaveAsync()
	// send out notifications
	.ContinueWith( task => {
		// do nothing if failed
		if( task.IsFaulted )
			return null;

		return NotificationService.SendOutNewMessageNotifications( chat, newMessage );
	});
```
⇨ ⇨ ⇨
```
return UnitOfWork.SaveAsync()
	// send out notifications
	.ContinueWith {
		// do nothing if failed
		if( value.IsFaulted )
			return null;

		return NotificationService.SendOutNewMessageNotifications( chat, newMessage );
	};
```


## EF Core
**Examples:**
1. DbContext model builders
2. Strategy executions

```
// 1: model builders
modelBuilder.Entity<Account>( account => {
	account.HasKey( _ => _.ID );
	account.HasAlternateKey( _ => _.ClientID );
	account.Property( _ => _.ClientID )
		.ValueGeneratedOnAdd();
	account.HasIndex( _ => _.ClientID )
		.IsUnique();
	account.Ignore( _ => _.HasPhoneNumber );
	account.Property( _ => _.PhoneNumber )
		.HasMaxLength( ColumnLength_CanonicalPhoneNumber )
		.HasConversion(
			// TODO(arch): Should we also serialize the user-friendly number into here? Might be needed later when we allow users to change their account phone numbers (unless always want clients to do this)
			model => (model != null) ? model.CanonicalNumber : null,
			raw => raw.NullOr( _ => new PhoneNumber( _, null )));
	account.HasIndex( _ => _.PhoneNumber )
		.IsUnique();
});
```
⇨ ⇨ ⇨
```
modelBuilder.Entity<Account> {
	_.HasKey( _.ID );
	_.HasAlternateKey( _.ClientID );
	_.Property( _.ClientID )	.ValueGeneratedOnAdd();
	_.HasIndex( _.ClientID ).IsUnique();
	_.Ignore( _.HasPhoneNumber );
	_.Property( _.PhoneNumber )
		.HasMaxLength( ColumnLength_CanonicalPhoneNumber )
		.HasConversion(
			// TODO(arch): Should we also serialize the user-friendly number into here? Might be needed later when we allow users to change their account phone numbers (unless always want clients to do this)
			model => (model != null) ? model.CanonicalNumber : null,
			raw => raw.NullOr( _ => new PhoneNumber( _, null )));
	_.HasIndex( _.PhoneNumber )
		.IsUnique();
};
```

```
// 2: execution strategy
ExecutionStrategy.ExecuteAsync( async () => {
	await using IDbContextTransaction transaction = await BeginTransaction();

	return await operation( transaction );
});
```
⇨ ⇨ ⇨
```
ExecutionStrategy.ExecuteAsync() async {
	await using IDbContextTransaction transaction = await BeginTransaction();

	return await operation( transaction );
};
```


## AutoMapper (not MS, but super popular/useful and would greatly benefit)
**Examples:**
1. Single mappings
2. Chained mappings
```
// 1: simple mapping
config.CreateMap<PlaceWhere, WhereDBO>()
	.ForMember( _ => _.Type, mapping => mapping.MapFrom( _ => WhereType.Place ));
```
⇨ ⇨ ⇨
```
config.CreateMap<PlaceWhere, WhereDBO>()
	.ForMember( _.Type ){
		value.MapFrom{ WhereType.Place }
	};
```

```
// 2: chained mappings
config.CreateMap<HappeningMemberState, HappeningMemberStateDBO>()
	.ForMember( _ => _.MemberRef, mapping => mapping.MapFrom( _ => _.Member ))
	.ForMember( _ => _.InvitedBy, mapping => mapping.Ignore() ) // NOTE: This is mapped at a higher level instead (ie: by Plan entity)
	.ForMember( _ => _.Computed_UnreadPlanUpdates, mapping => mapping.MapFrom( _ => _.Computed_UnreadHappeningUpdates ));
```
⇨ ⇨ ⇨
```
config.CreateMap<HappeningMemberState, HappeningMemberStateDBO>()
	.ForMember( _.MemberRef ){ _.MapFrom{ _.Member }}
	.ForMember( _.InvitedBy ){ _.Ignore() }
	.ForMember( _.Computed_UnreadPlanUpdates ){ _.MapFrom{ _.Computed_UnreadHappeningUpdates }};
```

**Reply by bondsbw** [reply]
Are the curly braces necessary for expressions?  Like with `Select`?

**Reply by HaloFour** [reply]
This implies some sigil to represent the lambda parameter as well, e.g. `_`.

To be honest, I don't really find these examples particularly compelling, but maybe that's just me.  When you chain off of the lambda it just feels muddy what you're actually chaining on.  I've seen better examples from Scala and Kotlin.

**Reply by CyrusNajmabadi** [reply]
My honest take is that the 'after' code looks quote confusing.

**Comment by Simplifier** [comment]
Is this proposal abandoned? Kotlin's lamdba syntax is so elegant and neat, it would be awesome to see it in c#

**Reply by CyrusNajmabadi** [reply]
@Simplifier No.  This is not abandoned.

**Comment by marchy** [comment]
Please. Do this. Already.

Biggest missing feature in C# by a long-shot.

**Comment by Richiban** [comment]
I love this feature (I use it mostly in Kotlin but Swift and Scala have it too, amongst others) and I think C# would benefit greatly from its introduction.

Even better would be if _somehow_ object initialiser expressions could be retconned into being trailing lambdas (inlined, of course) so that normal C# would be supported inside the expression:

```cs
var p = new Point 
{
	X = 4;

	if (someBool) 
	{
		Y = 5;
	}
}
```

This is a bit of a pain point around `init` properties where the property can only be set in the object initialiser but there's currently no way in C# to conditionally set a property.
